lib/trie/record/recorder.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package record
+
+// Recorder records the list of nodes found by Lookup.Find
+type Recorder struct {
+	nodes []Node
+}
+
+// NewRecorder creates a new recorder.
+func NewRecorder() *Recorder {
+	return &Recorder{}
+}
+
+// Record appends a node to the list of visited nodes.
+func (r *Recorder) Record(hash, rawData []byte) {
+	r.nodes = append(r.nodes, Node{RawData: rawData, Hash: hash})
+}
+
+// Next returns the first node in the recorded list
+// and removes it (shift operation).
+func (r *Recorder) Next() (node *Node) {
```

Comment:
taking advantage of this improvements I would like to suggest changing the behavior of the `Next` function, instead of shifting the data we could act as an iterator and keep track of an internal index that will be incremented each time the `Next` function is called and in the `IsEmpty` just return if the internal index is equal the `len(data)`

Reply:
That was a good idea, but I ended up making [it fast and dumb since we don't need all the fanciness](https://github.com/ChainSafe/gossamer/pull/2082/commits/ba6cc4010f8c61c535cadcfca7ee50715084dfaa)

Reply:
!it got even better!!